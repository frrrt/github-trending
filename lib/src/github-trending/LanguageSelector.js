import React from 'react';
import PropTypes from 'prop-types';

import Autocomplete from '@material-ui/lab/Autocomplete';
import { TextField, makeStyles } from '@material-ui/core';

const filterAvailableLanguages = currentLanguages => currentLanguages
    .sort()
    // Filter duplicate entries, from stackoverflow
    .filter((item, pos, ary) => !pos || item !== ary[pos - 1]);

const useStyles = makeStyles(() => ({
    root: {
        // width: 200 ensures that layout works on iPhone5 sized screens
        width: 200,
    },
}));

const LanguageSelector = ({ onChange, currentLanguages }) => {
    const classes = useStyles();

    return (
        <Autocomplete
            id="language-selector"
            options={filterAvailableLanguages(currentLanguages)}
            getOptionLabel={option => option}
            className={classes.root}
            size="small"
            selectOnFocus
            clearOnBlur
            handleHomeEndKeys
            // eslint-disable-next-line react/jsx-props-no-spreading
            renderInput={params => <TextField {...params} label="Language" variant="outlined" />}
            onChange={(_, newValue) => onChange(newValue)}
        />
    );
};

LanguageSelector.propTypes = {
    onChange: PropTypes.func.isRequired,
    currentLanguages: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default LanguageSelector;
