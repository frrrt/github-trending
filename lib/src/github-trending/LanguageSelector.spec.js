import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import LanguageSelector from './LanguageSelector';

const baseProps = {
    currentLanguages: ['JavaScript', 'Java', 'PHP', 'TypeScript', 'NonExistingLanguage'],
    onChange: jest.fn(),
};

describe('LanguageSelector', () => {
    it('is rendered correctly', () => {
        expect(render(<LanguageSelector {...baseProps} />).container).toMatchSnapshot();
    });

    it('select language', () => {
        render(<LanguageSelector {...baseProps} />);

        fireEvent.change(screen.getByRole('textbox'), { target: { value: 'Java' } });
        fireEvent.keyDown(screen.getByRole('textbox'), { keyCode: 13 });

        fireEvent.click(screen.getAllByRole('option')[1]);
        expect(baseProps.onChange.mock.calls.length).toEqual(1);
        expect(baseProps.onChange.mock.calls[0][0]).toEqual('JavaScript');
    });
});
