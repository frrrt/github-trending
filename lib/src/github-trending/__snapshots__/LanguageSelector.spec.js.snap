// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector is rendered correctly 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root makeStyles-root-1 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="language-selector"
        id="language-selector-label"
      >
        Language
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="language-selector"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Language
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
